rasterio/_err.pyx

```diff
@@ -185,13 +189,20 @@ cdef int exc_wrap(int retval) except -1:
     return retval
 
 
-cdef void chaining_error_handler(CPLErr err_class, int err_no, const char* msg) with gil:
-    msg_b = msg
-    message = msg_b.decode("utf-8")
-    exception = exception_map.get(err_no, CPLE_BaseError)(err_class, err_no, message)
-    error_stack = <object>CPLGetErrorHandlerUserData()
+cdef void chaining_error_handler(
+    CPLErr err_class,
+    int err_no,
+    const char* msg
+) noexcept with gil:
+    global chained_error_stack
     if err_class == 3:
-        error_stack.append(exception)
+        stack = chained_error_stack.get()
+        msg_b = msg
+        message = msg_b.decode("utf-8")
+        stack.append(
+            exception_map.get(err_no, CPLE_BaseError)(err_class, err_no, message),
+        )
+        chained_error_stack.set(stack)
```

Comment:
New code does nothing in the debug or warning cases. Only if we have a class 3 error.